common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py

```diff
@@ -2091,3 +2096,123 @@ def test_import_edit_import(self, default):
                 with self.store.branch_setting(ModuleStoreEnum.Branch.published_only, course_id):
                     published_vertical = self.store.get_item(vertical_loc)
                 self.assertEqual(draft_vertical.display_name, published_vertical.display_name)
+
+    def test_course_publish_signal_firing_split(self):
+        with MongoContentstoreBuilder().build() as contentstore:
+            self.store = MixedModuleStore(
+                contentstore=contentstore,
+                create_modulestore_instance=create_modulestore_instance,
+                mappings={},
+                signal_handler=SignalHandler(MixedModuleStore),
+                **self.OPTIONS
+            )
+            self.addCleanup(self.store.close_all_connections)
+
+            with self.store.default_store(ModuleStoreEnum.Type.split):
+                self.assertIsNotNone(self.store.thread_cache.default_store.signal_handler)
+
+                # Course creation should fire the signal
+                with mock_signal_receiver(SignalHandler.course_published) as receiver:
+                    self.assertEqual(receiver.call_count, 0)
+                    course = self.store.create_course('org_x', 'course_y', 'run_z', self.user_id,
+                                                      skip_auto_publish=True)
+                    self.assertEqual(receiver.call_count, 1)
+
+                    course_key = course.id
+
+                    # Test non-draftable block types. The block should be published with every change.
+                    categories = DIRECT_ONLY_CATEGORIES
+                    for block_type in categories:
+                        log.debug('Testing with block type %s', block_type)
+                        receiver.reset_mock()
+                        block = self.store.create_item(self.user_id, course_key, block_type)
+                        self.assertEqual(receiver.call_count, 1)
+
+                        block.display_name = block_type
+                        self.store.update_item(block, self.user_id)
+                        self.assertEqual(receiver.call_count, 2)
+
+                        self.store.publish(block.location, self.user_id)
+                        self.assertEqual(receiver.call_count, 3)
+
+                    # Test a draftable block type. Although blocks of this type are only published when the publish
+                    # method is explicitly called, our implementation will still emit the signal after each update.
```

Comment:
I discussed this with @ormsbee and we both are willing to accept the performance penalty. The first firing won't do much since, at that point, the course only has one node. Obviously, given that this is an asynchronous implementation, that may not always be the case.

I looked into your proposed solution. It won't work in this case because the `update_item` always has an active bulk operation. Ideally, we would disable signal emission when starting the import and re-enable/emit the signals at the end of the import. I just took a stab at this, and doing so will require a bit of argument passing between the `MixedModulestore` and the underlying modulestores. I'd prefer to hold off on this work until we have a better design in place that, hopefully, avoids adding too much Django-specific code in the modulestores.